### PR TITLE
Chapter 5

### PR DESCRIPTION
**Pull request by yossefkatri**
finished 5

### CODE REVIEW

File: src/main/java/encriptionAlgorithms/basicAlgorithms/ShiftMultiplyEncryption.java [hunk 1]
```diff
@@ -12,6 +12,12 @@
     private ShiftMultiplyEncryption() {
         UPPER_LIMIT = 9999;
     }
+
+    @Override
+    public String toString() {
+        return "{ShiftMultiplyEncryption}";
```
Comment (Shoval-Cohen) [inline src/main/java/encriptionAlgorithms/basicAlgorithms/ShiftMultiplyEncryption.java]: Change it at the parent class to be class name

File: src/main/java/encriptionAlgorithms/basicAlgorithms/ShiftMultiplyEncryption.java [hunk 2]
```diff
@@ -46,9 +52,9 @@ private int generatePrimeNum(int intKey) {
     }
     // encrypt the plaintext with the key and return the ciphertext
     @Override
-    public char encryptChar(char plaintext, IKey key) {
+    public char encryptChar(char plaintext, IKey<Integer> key) {
         //limited my key to 0-10
```
Comment (Shoval-Cohen) [inline src/main/java/encriptionAlgorithms/basicAlgorithms/ShiftMultiplyEncryption.java]: There is a log about this limitation?
Someone that will need to investigate the behavior of the algorithm will understand it without reading the code?

File: src/main/java/encriptionAlgorithms/complexAlgorithm/DoubleEncryption.java [hunk 3]
```diff
@@ -6,32 +6,37 @@
...
 
 
-public class DoubleEncryption extends EncryptionAlgorithmImpl {
-    final IEncryptionAlgorithm encryptionAlgorithm;
+public class DoubleEncryption<T> extends EncryptionAlgorithmImpl<T> {
+    final IEncryptionAlgorithm<T> encryptionAlgorithm;
 
 
-    public DoubleEncryption(EncryptionAlgorithmImpl aEncryptionAlgorithm) {
+    public DoubleEncryption(EncryptionAlgorithmImpl<T> aEncryptionAlgorithm) {
```
Comment (Shoval-Cohen) [inline src/main/java/encriptionAlgorithms/complexAlgorithm/DoubleEncryption.java:16]: Why not call it encryptionAlgorithm

File: src/main/java/encriptionAlgorithms/complexAlgorithm/DoubleEncryption.java [hunk 4]
```diff
@@ -6,32 +6,37 @@
...
         encryptionAlgorithm = aEncryptionAlgorithm;
         numberOfKeys = 2 * aEncryptionAlgorithm.numberOfKeys;
     }
 
     @Override
-    public char encryptChar(char plainChar, IKey key) {
-        char cipherChar = encryptionAlgorithm.encryptChar(plainChar, ((DoubleKey) key).getKey1());
-        cipherChar = encryptionAlgorithm.encryptChar(cipherChar, ((DoubleKey) key).getKey2());
+    public char encryptChar(char plainChar, IKey<T> key) {
+        char cipherChar = encryptionAlgorithm.encryptChar(plainChar, ((DoubleKey<T>) key).getKey1());
```
Comment (Shoval-Cohen) [inline src/main/java/encriptionAlgorithms/complexAlgorithm/DoubleEncryption.java]: What if the casting will not work?
If someone put key that is not DoubleKey?

File: src/main/java/encriptionAlgorithms/complexAlgorithm/RepeatEncryption.java [hunk 5]
```diff
@@ -36,4 +36,9 @@ public char decryptChar(char cipherChar, IKey key) {
     public int getKeyStrength() {
         return times * encryptionAlgorithm.getKeyStrength();
     }
+
+    @Override
+    public String toString() {
+        return "RepeatEncryption{"+encryptionAlgorithm+"}";
```
Comment (Shoval-Cohen) [inline src/main/java/encriptionAlgorithms/complexAlgorithm/RepeatEncryption.java]: Add how many times you repeat it

File: src/main/java/encryptor.java [hunk 6]
```diff
@@ -1,67 +1,188 @@
...
+import java.util.List;
 import java.util.Scanner;
 
 public class encryptor {
-
-
     public static void main(String[] args) {
         Scanner scanner = new Scanner(System.in);
+        Logger logger = LogManager.getLogger(encryptor.class);
+        isDebug();
```
Comment (Shoval-Cohen) [inline src/main/java/encryptor.java]: ?

File: src/main/java/encryptor.java [hunk 7]
```diff
@@ -1,67 +1,188 @@
...
         while (true) {
-            System.out.println("menu: \n 1: encryption \n 2: decryption");
+            System.out.println("menu: \n 1: encryption file \n 2: decryption file \n 3: encryption directory \n 4: decryption directory");
             int input = scanner.nextInt();
+            logger.debug("choice: " + input);
             scanner.nextLine();
-            IEncryptionAlgorithm encryptionAlgorithm =new DoubleEncryption(new DoubleEncryption( new DoubleEncryption(XorEncryption.getInstance())));
-            FileEncryptor fileEncryptor = new FileEncryptor(encryptionAlgorithm);
+            IEncryptionAlgorithm<Integer> encryptionAlgorithm = new DoubleEncryption<>(new DoubleEncryption<>(new DoubleEncryption<>(XorEncryption.getInstance())));
```
Comment (Shoval-Cohen) [inline src/main/java/encryptor.java]: What's going on? LoL

File: src/main/java/encryptor.java [hunk 8]
```diff
@@ -1,67 +1,188 @@
...
+            logger.debug("encryption operation: " + encryptionAlgorithm);
+            FileEncryptor<Integer> fileEncryptor = new FileEncryptor<>(encryptionAlgorithm);
+            KeyManager keyManager = new KeyManager(encryptionAlgorithm);
+            new EncryptionLog4JLogger(fileEncryptor.getStateChangeSupport());
```
Comment (Shoval-Cohen) [inline src/main/java/encryptor.java:63]: You don't save this logger in variable

File: src/main/java/encryptor.java [hunk 9]
```diff
@@ -1,67 +1,188 @@
...
-                stringPath = stringPath.replaceAll("\"", "");
                 try {
-                Path path = Paths.get(stringPath);
-                //calculate the path
-                Path filesPath = path.getParent();
-                Path outputFilepath = FileStream.getFileName(path, "_encrypted");
+                    //get the file-path from the user
+                    System.out.println("Enter the path of your source file:");
+                    String stringPath = scanner.nextLine();
+                    stringPath = stringPath.replaceAll("\"", "");
```
Comment (Shoval-Cohen) [inline src/main/java/encryptor.java]: why?

File: src/main/java/encryptor.java [hunk 10]
```diff
@@ -1,67 +1,188 @@
...
-                    System.out.println("The locations of the key file and the encrypted file :" + filesPath + "\n");
+                    logger.debug("generate keys for encryption");
+                    IKey<Integer> key = keyManager.getKeys();
+                    logger.debug("input file path:" + path + " output file path: " + outputFilepath + " key content: " + key);
+                    fileEncryptor.encryptFile(path, outputFilepath, key);
+
+                    logger.debug("start saving key operation with path:" + filesPath + "and key: " + key);
+                    Path keyPath = keyManager.saveKey(filesPath, key);
+                    logger.info("save key file successfully in " + keyPath);
                 } catch (Exception e) {
```
Comment (Shoval-Cohen) [inline src/main/java/encryptor.java]: Handle all exception types

File: src/main/java/encryptor.java [hunk 11]
```diff
@@ -1,67 +1,188 @@
...
+            logger.debug("choice: " + input);
             scanner.nextLine();
-            IEncryptionAlgorithm encryptionAlgorithm =new DoubleEncryption(new DoubleEncryption( new DoubleEncryption(XorEncryption.getInstance())));
-            FileEncryptor fileEncryptor = new FileEncryptor(encryptionAlgorithm);
+            IEncryptionAlgorithm<Integer> encryptionAlgorithm = new DoubleEncryption<>(new DoubleEncryption<>(new DoubleEncryption<>(XorEncryption.getInstance())));
+            logger.debug("encryption operation: " + encryptionAlgorithm);
+            FileEncryptor<Integer> fileEncryptor = new FileEncryptor<>(encryptionAlgorithm);
+            KeyManager keyManager = new KeyManager(encryptionAlgorithm);
+            new EncryptionLog4JLogger(fileEncryptor.getStateChangeSupport());
             if (input == 1) {
```
Comment (Shoval-Cohen) [inline src/main/java/encryptor.java]: use switch case

File: src/main/java/encriptionAlgorithms/basicAlgorithms/ShiftMultiplyEncryption.java [hunk 12]
```diff
@@ -12,6 +12,12 @@
     private ShiftMultiplyEncryption() {
         UPPER_LIMIT = 9999;
     }
+
```
Comment (Shoval-Cohen) [inline src/main/java/encriptionAlgorithms/basicAlgorithms/ShiftMultiplyEncryption.java:16]: Do ctrl +alt + L

File: src/main/java/events/DecryptionEndedEvent.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.nio.file.Path;
+import java.time.LocalDateTime;
+import java.util.EventObject;
+
+public class DecryptionEndedEvent extends EventObject {
+    private final EncryptionLogEventArgs encryptionLogEventArgs;
+
+    /**
+     * Constructs a prototypical Event.
```
Comment (Shoval-Cohen) [inline src/main/java/events/DecryptionEndedEvent.java]: Add informative information

File: src/main/java/events/EncryptionLogEventArgs.java [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+import java.nio.file.Path;
+import java.time.LocalDateTime;
+
+public class EncryptionLogEventArgs {
+    public LocalDateTime Time;
+    public String encryptionAlgorithmName;
+    public Path decrypted;
+    public Path encrypted;
+    public String key;
+    public Boolean isFile;
```
Comment (Shoval-Cohen) [inline src/main/java/events/EncryptionLogEventArgs.java]: use primitive type

File: src/main/java/ilisteners/IDecryptionEndedListener.java [hunk 15]
```diff
@@ -0,0 +1,7 @@
+package ilisteners;
```
Comment (Shoval-Cohen) [inline src/main/java/ilisteners/IDecryptionEndedListener.java]: i?

File: src/main/java/keys/DoubleKey.java [hunk 16]
```diff
@@ -1,24 +1,29 @@
 package keys;
 
-public class DoubleKey implements IKey {
-    final IKey key1;
-    final IKey key2;
+public class DoubleKey<T> implements IKey<T> {
```
Comment (Shoval-Cohen) [inline src/main/java/keys/DoubleKey.java]: Double key must be the same type?

**Review by Shoval-Cohen — CHANGES_REQUESTED** [review]

File: src/main/java/multiThreading/AsyncDirectoryProcessor.java [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.concurrent.Future;
+
+public class AsyncDirectoryProcessor<T> implements IDirectoryProcessor<T>{
+    final private FileEncryptor<T> fileEncryptor;
+
+    public AsyncDirectoryProcessor(FileEncryptor<T> fileEncryptor) {
+        this.fileEncryptor = fileEncryptor;
+    }
+
+    private LocalDateTime getPeriod(LocalDateTime startTime) {
```
Comment (Shoval-Cohen) [inline src/main/java/multiThreading/AsyncDirectoryProcessor.java:31]: Read about java Duration object

File: src/main/java/multiThreading/AsyncDirectoryProcessor.java [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+        LocalDateTime endTime = LocalDateTime.now();
+        long mills = ChronoUnit.MILLIS.between(startTime, endTime);
+        return Instant.ofEpochMilli(mills).atZone(ZoneId.of("UTC")).toLocalDateTime();
+    }
+
+    @Override
+    public void encryptDirectory(Path originalDirPath, Path outputDirPath, IKey<T> key) throws Exception {
+
```
Comment (Shoval-Cohen) [inline src/main/java/multiThreading/AsyncDirectoryProcessor.java:39]: Empty line

File: src/main/java/multiThreading/AsyncDirectoryProcessor.java [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+        IEncryptionAlgorithm<T> encryptionAlgorithm = fileEncryptor.getEncryptionAlgorithm();
+        LocalDateTime startTime = LocalDateTime.now();
+        fileEncryptor.getStateChangeSupport().notifyEncryptionStartedListeners(this, startTime, encryptionAlgorithm.toString(), outputDirPath, originalDirPath, key.toString(), false);
```
Comment (Shoval-Cohen) [inline src/main/java/multiThreading/AsyncDirectoryProcessor.java]: break to lines

File: src/main/java/utils/StateChangeSupport.java [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+    private final List<IDecryptionEndedListener> decryptionEndedListeners = new ArrayList<>();
+    private final List<IEncryptionEndedListener> encryptionEndedListeners = new ArrayList<>();
+    private final List<IEncryptionStartedListener> encryptionStartedListeners = new ArrayList<>();
+
+    public  void addDecryptionStartedListener(IDecryptionStartedListener decryptionStartedListener) {
+        decryptionStartedListeners.add(decryptionStartedListener);
+    }
+
+    public  void notifyDecryptionStartedListeners(Object source, LocalDateTime startTime, String encryptionAlgorithmName, Path decryptedFilePath, Path encryptedFilePath, String key, Boolean isFile) {
+        for (IDecryptionStartedListener listener :
```
Comment (Shoval-Cohen) [inline src/main/java/utils/StateChangeSupport.java]: use Java Streams as you learned in the java8 tutorial

File: src/main/java/utils/FileStream.java [hunk 21]
```diff
@@ -13,6 +13,12 @@
 
 public final class FileStream {
 
+    private static void repeatWrite(FileWriter fileWriter, String template) throws IOException {
+        for (int i = 0; i< (long) 262144; ++i)
```
Comment (Shoval-Cohen) [inline src/main/java/utils/FileStream.java]: ????

File: src/main/java/utils/FileStream.java [hunk 22]
```diff
@@ -36,14 +47,19 @@ public static File createFile(Path fileName) throws IOException {
...
        }
        catch (IOException ex) {
            throw new IOException("The function can't succeed to create file. \n File: "+fileName.getFileName() +"\n Error: "+ex.getMessage());
        }
     }
+    //create directory by his name
+    public static void createDirectory(Path dirPath) throws IOException {
+        File theDir = new File(dirPath.toString());
+        if(!(theDir.mkdirs() | theDir.exists())) {
+            throw new IOException("The system cannot create the directory");
```
Comment (Shoval-Cohen) [inline src/main/java/utils/FileStream.java:61]: ||

File: src/main/java/utils/FileStream.java [hunk 23]
```diff
@@ -81,4 +104,18 @@ public static void saveData(File file, String message) throws IOException {
         }
 
     }
+
+    public static void generateBigFiles(Path dir, int numOfFile) throws IOException {
```
Comment (Shoval-Cohen) [inline src/main/java/utils/FileStream.java:110]: Add doc

File: src/main/java/utils/KeyManager.java [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+            return null;
+        }
+        IKey<Integer> key1;
+        key1 = buildKey(keys.subList(0, keys.size() / 2));
+
+        IKey<Integer> key2;
+        key2 = buildKey(keys.subList(keys.size() / 2, keys.size()));
+
+        return new DoubleKey<>(key1, key2);
+    }
```
Comment (Shoval-Cohen) [inline src/main/java/utils/KeyManager.java:43]: add empty line

File: src/main/java/utils/KeyManager.java [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+
+public class KeyManager {
```
Comment (Shoval-Cohen) [inline src/main/java/utils/KeyManager.java:20]: Ctrl + Alt + L

File: src/main/java/utils/KeyManager.java [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+        key2 = buildKey(keys.subList(keys.size() / 2, keys.size()));
+
+        return new DoubleKey<>(key1, key2);
+    }
+    public  void checkKeys(List<Integer> keys) throws InvalidEncryptionKeyException {
+        int numberOfKeys = ((EncryptionAlgorithmImpl<?>) encryptionAlgorithm).getNumberOfKeys();
+        if (numberOfKeys != keys.size()) {
+            throw new InvalidEncryptionKeyException("Number of keys: " + keys.size() + "  \nExpected number of key: " + numberOfKeys);
+        }
+        for (Integer key : keys) {
```
Comment (Shoval-Cohen) [inline src/main/java/utils/KeyManager.java]: use streams
Adi makeup was here

File: src/main/java/utils/KeyManager.java [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+            if (key < 0 || key.toString().length() > encryptionAlgorithm.getKeyStrength())
+            {
+                throw new InvalidEncryptionKeyException(key, "The key should be between 0 to " + encryptionAlgorithm.getKeyStrength() +"digits");
+            }
+        }
+    }
```
Comment (Shoval-Cohen) [inline src/main/java/utils/KeyManager.java:54]: Empty line

File: src/main/java/utils/KeyManager.java [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+    public  IKey<Integer> getKeys()  {
+        //get the necessary number of the keys
+        int numKeys = ((EncryptionAlgorithmImpl<?>) encryptionAlgorithm).getNumberOfKeys();
+        List<Integer> keys = new ArrayList<>();
+        for (int i = 0; i < numKeys; ++i) {
```
Comment (Shoval-Cohen) [inline src/main/java/utils/KeyManager.java]: Google: create random int list java

File: src/main/java/utils/KeyManager.java [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+            //generate the key
+            Random randomizer = new Random();
+            int tempKey = randomizer.nextInt(UPPER_LIMIT);
+            keys.add(tempKey);
+        }
+        return buildKey(keys);
+    }
```
Comment (Shoval-Cohen) [inline src/main/java/utils/KeyManager.java:63]: empty line

File: src/main/java/utils/KeyManager.java [hunk 30]
```diff
@@ -0,0 +1,70 @@
...
+    public  Path saveKey(Path outputPath, IKey<Integer> key) throws IOException {
+        Path keyPath = Paths.get(outputPath.toString(), "key.txt");
```
Comment (Shoval-Cohen) [inline src/main/java/utils/KeyManager.java:66]: It wil override already saved key
How do you handle it?

File: src/main/java/utils/KeyManager.java [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+            throw new InvalidEncryptionKeyException("Number of keys: " + keys.size() + "  \nExpected number of key: " + numberOfKeys);
+        }
+        for (Integer key : keys) {
+            if (key < 0 || key.toString().length() > encryptionAlgorithm.getKeyStrength())
+            {
+                throw new InvalidEncryptionKeyException(key, "The key should be between 0 to " + encryptionAlgorithm.getKeyStrength() +"digits");
+            }
+        }
+    }
+    public  IKey<Integer> getKeys()  {
```
Comment (Shoval-Cohen) [inline src/main/java/utils/KeyManager.java]: Rename to generateKeys

File: src/main/java/encryptor.java [hunk 32]
```diff
@@ -1,67 +1,188 @@
...
+
+                    //calculate the path
+                    Path filesPath = path.getParent();
+                    Path outputFilepath = FileStream.getFileName(path, "_encrypted");
 
-                    fileEncryptor.encryptFile(path,outputFilepath,filesPath);
-                    System.out.println("The locations of the key file and the encrypted file :" + filesPath + "\n");
+                    logger.debug("generate keys for encryption");
+                    IKey<Integer> key = keyManager.getKeys();
+                    logger.debug("input file path:" + path + " output file path: " + outputFilepath + " key content: " + key);
```
Comment (Shoval-Cohen) [inline src/main/java/encryptor.java]: use the logger built-in

**Review by Shoval-Cohen — CHANGES_REQUESTED** [review]
Ctrl + Alt + L

File: src/main/java/encryptor.java (hunk 32 above)
Comment (yossefkatri) [reply src/main/java/encryptor.java]: done

File: src/main/java/utils/FileStream.java (hunk 21 above)
Comment (yossefkatri) [reply src/main/java/utils/FileStream.java]: the number of time you should repeat to get about 3MB file.

**Comment by yossefkatri** [comment]
done

**Pull request merged**